Zend/zend_compile.c

```diff
@@ -8050,9 +8121,37 @@ void zend_compile_top_stmt(zend_ast *ast) /* {{{ */
 
 	if (ast->kind == ZEND_AST_STMT_LIST) {
 		zend_ast_list *list = zend_ast_get_list(ast);
-		uint32_t i;
-		for (i = 0; i < list->children; ++i) {
-			zend_compile_top_stmt(list->child[i]);
+		zend_ast **begin = list->child;
```

Comment:
As already mentioned, this code is not going to handle namespaces correctly, especially when using braced `namespace Foo { ... }` form. You'll want classes that are part of multiple braced (or unbraced, but braced is the more complex case) to be part of the same variance region.

Reply:
I think I finally understand what you are saying:

```php
namespace NS1 {
  class A implements B {}
}

namespace NS2 {
  class X implements Y {}
}
```

The `A` should be verified after `X` is declared. Right?

Reply:
Yes.